guardian/mobile-n10n
# Set up GuCDK stack for Registration and deploy stage-specific templates

## What does this change?

We would like to use the new constructs for error budget alerting (https://github.com/guardian/cdk/pull/1576) so that we receive warnings if we are at risk of missing SLO(s) for the Registration service.

This PR imports the Registration YAML via GuCDK in order to configure these alerts, as doing so will make it easier[^1] to reference the correct metrics in TypeScript code and also simplifies deploying the error budget alarm configuration via Riff-Raff. Finally, the work here is the [first step in migrating the Registration app to GuCDK](https://github.com/guardian/cdk/blob/main/docs/migration-guide.md#introduce-cdk-to-the-repository-with-cicd)[^2], so we will need to do it at some point anyway! 

cc @guardian/devx-reliability 

## How to test

This shouldn't change any behaviour - I've [deployed to `CODE`](https://riffraff.gutools.co.uk/deployment/view/4837ee12-d059-4ef3-855b-dab753eaf464) and can see that resources are being updated (but, crucially, not deleted/replaced!) by this change:

![image](https://user-images.githubusercontent.com/19384074/203331462-6a9ab4e1-e82c-47ad-ace4-2c0a2e5e5435.png)

The updates are just GuCDK automatically tagging resources e.g.

![image](https://user-images.githubusercontent.com/19384074/203331729-4f3a1806-f841-4561-8cf7-8324c3c0b7bf.png)

## How can we measure success?

* Everything should continue working as before, but we'll be able to make use of `GuErrorBudgetAlarmExperimental` more easily in a future PR.
* There is slightly less work to do when we decide to migrate Registration to GuCDK.

## Have we considered potential risks?

We are moving to two separate infrastructure definition files here (one for `CODE` and one for `PROD`) so there is a risk that I've made a typo and wired things up incorrectly (although I will test in `CODE` there are now some `PROD`-specific errors that would be missed via testing, so a careful review would be much appreciated!).

Lastly, operating in this partially-migrated state means that the process for updating the infrastructure for Registration could be slightly more confusing than before (you need to update the YAML and then update the snapshot tests); I've tried to mitigate this with comments to explain the new process.

[^1]: That said, it is not strictly necessary to do this now - we could just deploy a standalone alerting stack and reference the SLI metrics by hardcoding the names as strings. If you'd prefer not to import the current template let me know and I'll take a different approach.
[^2]: It's a slightly amended version of the steps described in the GuCDK migration docs, since this repo already includes the GuCDK library; the end result is the same. In other words, when you decide to fully migrate Registration to GuCDK, you'll be able to start [here](https://github.com/guardian/cdk/blob/main/docs/migration-guide-ec2.md).
